lib/vbt/provider.ex

```diff
@@ -0,0 +1,289 @@
+
+  Default options are considered in the following order:
+
+  1. `:test` (if mix env is `:test`)
+  2. `:dev` (if mix env is either `:dev` or `:test`)
+  3. `:default`
+
+  For example, if `:test` and `:default` options are given, the `:test` value will be used as a
+  default in `:test` env, while `:default` will be used in all other envs.
+
+  When you invoke the generated functions, values will be retrieved from the external storage
+  (e.g. OS env). If some value is not available, a default value will be used (if provided). The
+  values are then casted according to the parameter type.
+
+  Each default can be a constant, but it can also be an expression, which is evaluated at runtime.
+  For example:
+
+      defmodule MySystem.Config do
+        use VBT.Provider,
+          source: VBT.Provider.SystemEnv,
+          params: [
+            # db_name/0 will be invoked when you try to retrieve this parameter (or all parameters)
+            {:db_name, dev: db_name()},
+            # ...
+          ]
+
+        defp db_name(), do: #...
+      end
+
+  It's worth noting that `VBT.Provider` performs compile-time purging of needless defaults. When you
+  compile the code in `:prod`, `:dev` and `:test` defaults will not be included in the binaries.
+  Consequently, any private function invoked only in dev/test will also not be invoked, and you'll
+  get a compiler warning when compiling the code in prod. To eliminate such warnings, you can
+  conditionally define the function only in required mix env, by moving the function definition
+  under an `if Mix.env() == ` conditional.
+
+  ## Generating template
+
+  The config module will contain the `template/0` function which generates the configuration
+  template. To print the template to stdout, you can invoke:
+
+      MIX_ENV=prod mix compile
+      MIX_ENV=prod mix run --no-start -e 'IO.puts(MySystem.Config.template())'
+
+  ## Lower level API
+
+  The foundational retrieval functionality is available via functions of this module, such as
+  `fetch_all/2`, or `fetch_one/2`. These functions are a lower level plumbing API which is less
+  convenient to use, but more flexible. Most of the time the `use`-based interface will serve you
+  better, but if you have some more involved needs which are not covered by that, you can reach
+  for these functions.
+  """
+
+  alias Ecto.Changeset
+
+  @type source :: module
+  @type params :: %{param_name => param_spec}
+  @type param_name :: atom
+  @type param_spec :: %{type: type, default: value}
+  @type type :: :string | :integer | :float | :boolean
+  @type value :: String.t() | number | boolean | nil
+  @type data :: %{param_name => value}
+
+  # ------------------------------------------------------------------------
+  # API
+  # ------------------------------------------------------------------------
+
+  @doc "Retrieves all params according to the given specification."
+  @spec fetch_all(source, params) :: {:ok, data} | {:error, [String.t()]}
```

Comment:
What possible sources could we have?

Reply:
Good question! Possible other sources could include json/ini files, etcd, hashicorp vault & such. AFAIK, currently we only need `System.get_env`, so this is a bit of an upfront overdesign. However, I had a couple of reasons for doing it upfront:

1. Since the code is heavily macro based, I wanted to make some possible extensions easier, so that you don't have to completely understand the macro part. This is also why lower level plumbing (such as `fetch_all(source, params)`) is separate from the macro code.
2. I think that we should opensource this stuff and publish to hex, so I wanted to immediately make it more flexible.
3. I think the code is slightly more explicit and self-explanatory when we have something like `use Provider, source: Provider.SystemEnv, ...`. It should be a bit more obvious to the reader that stuff is retrieved from OS env, without requiring reading the code of the dep.